File: libmango/py/libmango.py
import io, re, time, signal, os, random
from serialiser import *
from dataflow import *
from transport import *
from interface import *
from error import *
import traceback
from inspect import getframeinfo, stack

class m_node: 
    def __init__(self,debug=False):
        self.version = "0.1"
        self.debug = debug
        self.context = zmq.Context()
        self.poller = zmq.Poller()
        self.serialiser = m_serialiser(self.version)
        self.interface = m_if(default_handler=self.default)
        self.dataflows = {}
        self.node_id = os.getenv('MANGO_COOKIE')
        self.interface.add_interface({'heartbeat':self.heartbeat,'exit':self.end})
        self.flags = {}
        self.server = os.getenv('MANGO_SERVER',None)
        if not self.server is None:
            self.local_gateway = m_ZMQ_transport(self.server,self.context,self.poller,''.join(random.choice('ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789') for i in range(10)))
            s = self.local_gateway.socket
            self.dataflow = m_dataflow(self.interface,self.local_gateway,self.serialiser,self.dispatch,self.handle_error)
            self.dataflows[s] = self.dataflow
            self.poller.register(s,zmq.POLLIN)

    def add_socket(self, sock, recv_cb, err_cb):
        dataflow = lambda: None
        dataflow.recv = recv_cb
        dataflow.die = err_cb
        self.poller.register(sock, zmq.POLLIN)
        self.dataflows[sock] = dataflow
            
    def ready(self):
        self.m_send('alive',{})
            
    def dispatch(self,header,args):
        self.debug_print("DISPATCH",header,args)
        try:
           result = self.interface.get_function(header['command'])(args)
           if not result is None:
               self.m_send(result[0],result[1],mid=header.get('mid',None))
        except Exception as exc:
           self.handle_error("unknown",traceback.format_exc())

    def heartbeat(self,args):
        self.mc_send('alive',{})
        
    def end(self,args):
        exit(0)
            
    def default(self,args):
        self.debug_print("UNHANDLED",header,args)

    def handle_error(self,src,err):
        self.debug_print('OOPS',src,err)
        self.mc_send('error',{'source':src,'message':err})

    def make_header(self,command,mid=None):
        header = {
            'command':command,
            'mid':mid if not mid is None else ''.join(random.choice('ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789') for i in range(10)),
            'cookie':self.node_id,
            'format':'json'}
        self.debug_print("H",header)
        return header

    def m_send(self,name,msg,mid=None):
        self.debug_print('sending',msg)
        self.dataflow.send(self.make_header(name,mid=mid),msg)

    def debug_print(self,*args):
        if self.debug:
            caller = getframeinfo(stack()[1][0])
            print("[LIBMANGO.PY] [{}:{} {} DEBUG] ".format(caller.filename, caller.lineno, self.node_id),*args)
    
    def run(self,f=None):
        if not self.server is None:
            self.ready()
        while True:
            socks = dict(self.poller.poll(1000*1000))
            for s in socks:
                if socks[s] == zmq.POLLIN:
                    self.debug_print("RX",s,self.dataflows[s])
                    self.dataflows[s].recv()
                elif(socks[s] & zmq.POLLERR != 0):
                    self.debug_print('socket error',s,socks[s])
                    self.dataflows[s].die()
            if not f is None: 
                f()

File: libmango/js/libmango.js
zmq = require('zmq');
jsyaml = require('js-yaml');
fs = require('fs');

function MNode(debug){
    var self = this;
    this.version = "0.1";
    this.debug = debug;
    this.serialiser = new Serialiser(this.version);
    this.iface = new Interface();
    this.node_id = process.env['MANGO_ID'];
    this.ports = [];
    var server = process.env['MC_ADDR'];
    
    this.run = function(){
	var ifce = {};
	for(var i in this.iface.iface){
	    ifce[i] = JSON.parse(JSON.stringify(this.iface.iface[i]));
	    delete ifce[i]['handler'];
	}
	console.log("IF",ifce)
	self.m_send('hello',{'id':self.node_id,'if':ifce,'ports':self.ports},"mc")
    }
    
    this.dispatch = function(header,args){
	console.log("DISPATCH",header,args,self.iface.iface,self.iface.iface[header['command']]);
	try{
            result = self.iface.iface[header['command']]['handler'](header,args);
            if(result){
		self.m_send("reply", result, header['port'])
	    }
	} catch(e) {
	    console.log(e);
            self.handle_error(header['src_node'],e+"")
	}
    }
    
    this.handle_error = function(src,err){
	console.log('OOPS',src,err);
	self.m_send('error',{'source':src,'message':err},"mc");
    }

    this.reg = function(header,args){
	if(header.src_node == "mc") self.node_id = args["id"];
    }

    this.reply = function(header,args){
	console.log("REPLY",header,args);
    }

    this.heartbeat = function(header,args){
	self.m_send("alive",{},"mc");
    }

    this.make_header = function(command,src_port){
	if(!src_port) src_port = "stdio";
	return {'src_port':src_port, 'command':command};
    }

    this.m_send = function(command,msg,port){
	console.log('sending',command,msg,port)
	header = self.make_header(command,port)
	self.dataflow.send(header,msg)
    }
    
    this.iface.add_interface(process.env['NODE_PATH']+'/../node.yaml',{
        'reg':self.reg, 'reply':self.reply, 'heartbeat':self.heartbeat
    });
    this.local_gateway = new Transport(server);
    var s = this.local_gateway.socket;
    this.dataflow = new Dataflow(self.iface, self.local_gateway, self.serialiser, self.dispatch, self.handle_error);
    s.on('message',function(data){ self.dataflow.recv(data); });
}

function MError(message){
    this.message = message;
    this.name = "Mango Error";
}

function Interface(){
    var self = this;
    this.iface = {};
    
    this.add_interface = function(if_file,handlers){
	try {
	    var spec = jsyaml.safeLoad(fs.readFileSync(if_file, 'utf8'));
	    console.log(JSON.stringify(spec));
	    var name = spec.name;
	    var missing = [], extra = [];
	    for(var i in handlers)
		if(!(i in spec.inputs)) extra.push(i);
	    for(var i in spec.inputs)
		if(!(i in handlers)) missing.push(i);
	    if(extra.length > 0) throw new MError("Functions not in interface: "+extra.join(", "));
	    if(missing.length > 0) throw new MError("Functions not implemented: "+missing.join(", "));

	    self.iface[name] = {};
	    for(var i in handlers){
		self.iface[name][i] = spec.inputs[i] ? spec.inputs[i] : {};
		self.iface[name][i]['handler'] = handlers[i];
	    }
	} catch (e) {
	    console.log(e);
	    throw new MError("Failed to load interface");
	}
    }

    this.get_function = function(fn,ns){
	if(ns) return self.iface[ns][fn];
	for(var n in self.iface) if(fn in self.iface[n]) return self.iface[n][fn];
	throw new MError("Functions not implemented: "+ns+"."+fn);	
    }
    
    this.validate = function(fn){
	return fn in self.iface;
    }
}

function Transport(target){
    var self = this;
    this.target = target
    this.socket = zmq.socket("dealer");
    this.socket.connect(target);
    
    this.tx = function(data){
	self.socket.send(data);
    }

}

function Dataflow(iface,transport,serialiser,dispatch_cb,error_cb){
    var self = this;
    this.iface = iface;
    this.transport = transport;
    this.serialiser = serialiser;
    this.dispatch_cb = dispatch_cb;
    this.error_cb = error_cb;
    console.log(this.dispatch_cb);
    this.send = function(header,args){
	self.transport.tx(self.serialiser.serialise(header,args));
    }
    
    this.recv = function(data){
	try{
	    var m = self.serialiser.deserialise(data);
	    if(!self.iface.validate(m[0]["command"])) throw new MError("Unknown function");
	    console.log(self.dispatch_cb);
	    self.dispatch_cb(m[0],m[1]);
	} catch(e) {
	    console.log(e);
	    self.error_cb(e);
	}
    }

}

function Serialiser(version){
    var self = this;
    this.version = version;
    this.method = "json";
    
    this.make_preamble = function(){
	return "MANGO"+self.version+" json\n";
    }

    this.parse_preamble = function(data){
	console.log("GOT",data);
	var nl1 = data.indexOf('\n');
	var m = data.substring(0,nl1).match(/^MANGO([0-9.]*) ([^ ]*)$/);
	if(!m || m.length < 1) throw new MError("Preamble failed to parse");
	return [m[1],m[2],data.substring(nl1)];
    }

    this.serialise = function(header,args){
	return self.make_preamble()+JSON.stringify({"header":header,"args":args});
    }
    
    this.deserialise = function(data){
	data = data.toString('utf8');
	p = self.parse_preamble(data);
	var ver = p[0]; var method = p[1]; var message = p[2];
	if(ver != self.version) throw new MError("Version mismatch");
	if(method != self.method) throw new MError("Unsupported method");
	try{
	    var d = JSON.parse(message);
	    return [d['header'],d['args']]
	} catch (e) {
	    console.log(e);
	    throw new MError("Failed to parse message");
	}
    }
}

module.exports = MNode;
